### PR TITLE
allow usage of user-defined docker images in runpod

### PR DESCRIPTION
**Pull request by winglian**
Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/clouds/runpod.py [hunk 1]
```diff
@@ -183,10 +181,16 @@ def make_deploy_resources_variables(
         else:
             custom_resources = None
 
+        if r.extract_docker_image() is not None:
+            image_id = r.extract_docker_image()
+        else:
+            image_id = r.image_id
```
Comment (Michaelvll) [inline sky/clouds/runpod.py:187]: Let's use our kubernetes default image for runpod as well to speed up the provisioning:

```suggestion
        if r.image_id is None:
            # Use the same default image as kubernetes cloud for
            # the best performance.
            image_id = service_catalog.get_image_id_from_tag(
                'skypilot:gpu-ubuntu-2004', clouds='kubernetes')
        elif None in r.image_id:
            image_id = r.image_id[None]
        else:
            image_id = r.image_id[region.name]

        if image_id.startswith('docker:'):
            image_id = image_id[len('docker:'):]
```

File: sky/provision/runpod/instance.py [hunk 2]
```diff
@@ -83,7 +83,9 @@ def run_instances(region: str, cluster_name_on_cloud: str,
                 name=f'{cluster_name_on_cloud}-{node_type}',
                 instance_type=config.node_config['InstanceType'],
                 region=region,
-                disk_size=config.node_config['DiskSize'])
+                disk_size=config.node_config['DiskSize'],
+                image_name=config.node_config['ImageId'] or 'runpod/base:0.0.2',
```
Comment (Michaelvll) [inline sky/provision/runpod/instance.py:87]: With the handling of `image_id` in `runpod.py`, we do not need the `or` logic here.
```suggestion
                image_name=config.node_config['ImageId'],
```

File: sky/clouds/runpod.py [hunk 3]
```diff
@@ -35,8 +35,6 @@ class RunPod(clouds.Cloud):
              'supported yet on RunPod.'),
         clouds.CloudImplementationFeatures.IMAGE_ID:
             ('Specifying image ID is not supported on RunPod.'),
```
Comment (Michaelvll) [inline sky/clouds/runpod.py:37]: Let's remove this as well, similar to our handling for image in kubernetes
```suggestion
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding the support @winglian! We are asking our RunPod contact for the issue of using custom docker image fails to expose port 22 to public.

Did you encounter this issue while testing this PR?
```
sky launch -c test-runpod --image-id us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot-gpu:20240613
```
The command above will stuck at `Waiting for SSH connection`

File: sky/clouds/runpod.py [hunk 4]
```diff
@@ -279,3 +283,7 @@ def validate_region_zone(self, region: Optional[str], zone: Optional[str]):
         return service_catalog.validate_region_zone(region,
                                                     zone,
                                                     clouds='runpod')
+
+    @classmethod
+    def get_image_size(cls, image_id: str, region: Optional[str]) -> float:
+        return 0.0
```
Comment (Michaelvll) [inline sky/clouds/runpod.py:289]:
```suggestion
        # TODO: use 0.0 for now to allow all images. We should change this to
        # return the docker image size.
        return 0.0
```

File: sky/clouds/runpod.py (hunk 1 above)
Comment (winglian) [reply sky/clouds/runpod.py:187]: doesn't this duplicate some of the code from `extract_docker_image()` ?

**Comment by winglian** [comment]
> Did you encounter this issue while testing this PR?
> 
> ```
> sky launch -c test-runpod --image-id us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot-gpu:20240613
> ```
> 
I think the docker image might be a bit finicky. Does your skypilot image actually setup sshd? This works for me, but our image has sshd running. You don't need sshd if you can connect via the runpod proxy, but I'm not sure enough info is exposed to do that.

```bash
sky launch --cloud runpod -c test-runpod-v2 --image-id docker:winglian/axolotl-cloud-term:main-py3.11-cu121-2.3.0
```

**Comment by winglian** [comment]
I think longer term, we would need a way to capture the organization id so that we could build the correct user id to connect to ssh via the internal proxy.

File: sky/clouds/runpod.py (hunk 1 above)
Comment (Michaelvll) [reply sky/clouds/runpod.py:185]: Hi @winglian, it is slightly different, as we have to `r.image_id` is a dict and we have to handle it. Also, we would like to set the default value for `image_id`. An alternative way is to:
```suggestion
            if r.image_id is None:
                image_id = 'runpod/base:0.0.2'
            elif r.extract_docker_image() is not None:
                image_id = r.extract_docker_image()
            else:
                image_id = r.image_id[r.region]
```

**Comment by Michaelvll** [comment]
> I think the docker image might be a bit finicky. Does your skypilot image actually setup sshd? This works for me, but our image has sshd running. You don't need sshd if you can connect via the runpod proxy, but I'm not sure enough info is exposed to do that.

Thanks for pointing this out! We were trying to figure out a reason why RunPod does not expose the port correctly, and it turns out it is the ssh service issue. We just figured a way to install the ssh service in the container, would you like to adopt those changes to this PR as well (basically adding a start command to set up ssh, rsync and other necessary dependencies)? 
https://github.com/skypilot-org/skypilot/blob/ad3bb203d1f0aad1e1f84d005249ead54c8b770b/sky/provision/runpod/utils.py#L108-L147


Currently, we still have to figure out how to make it work with docker images that do not have `root` as the default user name, but I am good with having this PR merged first : )

**Comment by winglian** [comment]
> would you like to adopt those changes to this PR as well (basically adding a start command to set up ssh, rsync and other necessary dependencies)?

How about some sort of configuration option to set up sshd, etc instead of it being the default? For most people building images for runpod, sshd will already be configured with that, and it might break unexpectedlyto try to install I again.

**Comment by romilbhardwaj** [comment]
Nit - might be good to update our docs as well to something like:

```
.. note::

    Running docker containers is `not supported on RunPod <https://docs.runpod.io/references/faq#can-i-run-my-own-docker-daemon-on-runpod>`_. To use RunPod, either use your image :ref:`as a runtime environment <docker-containers-as-runtime-environments>` or use ``setup`` and ``run`` to configure your environment. See `GitHub issue <https://github.com/skypilot-org/skypilot/issues/3096#issuecomment-2150559797>`_ for more.

```

https://github.com/skypilot-org/skypilot/blob/994d35ae3bf0962e1d21e4db006c2ba1dddcaa0f/docs/source/examples/docker-containers.rst?plain=1#L11-L13

**Comment by Michaelvll** [comment]
> > would you like to adopt those changes to this PR as well (basically adding a start command to set up ssh, rsync and other necessary dependencies)?
> 
> How about some sort of configuration option to set up sshd, etc instead of it being the default? For most people building images for runpod, sshd will already be configured with that, and it might break unexpectedlyto try to install I again.

Based on our experience with people's custom docker image for kubernetes, these setup commands do not affect the docker images already with sshd setup, and it seems the scripts working fine with the image you shown above. How about we keep it by default for now to make it align with the behavior of k8s? ; )

**Review by Michaelvll — APPROVED** [review]
I am happy with merging this PR as is and getting those fixes in another PR : )

Thanks a lot for helping add this support @winglian!

**Comment by Michaelvll** [comment]
Tested:
- [ ] `sky launch --cloud runpod nvidia-smi`
- [ ] `sky launch --image-id winglian/axolotl:main-latest --cloud runpod nvidia-smi -c test-axolotl`

**Comment by winglian** [comment]
Thanks for getting this over the line @Michaelvll

**Pull request merged**